db/db_iter.cc

```diff
@@ -797,20 +801,33 @@ bool DBIter::FindValueForCurrentKey() {
       return false;
     }
 
+    if (!user_comparator_.EqualWithoutTimestamp(ikey.user_key,
+                                                saved_key_.GetUserKey())) {
+      // Found a smaller user key, thus we are done with current user key.
+      break;
+    }
+
     assert(ikey.user_key.size() >= timestamp_size_);
     Slice ts;
     if (timestamp_size_ > 0) {
       ts = Slice(ikey.user_key.data() + ikey.user_key.size() - timestamp_size_,
                  timestamp_size_);
     }
-    if (!IsVisible(ikey.sequence, ts) ||
-        !user_comparator_.EqualWithoutTimestamp(ikey.user_key,
-                                                saved_key_.GetUserKey())) {
+
+    bool visible = IsVisible(ikey.sequence, ts);
+    if (!visible &&
+        (timestamp_lb_ == nullptr ||
+         user_comparator_.CompareTimestamp(ts, *timestamp_ub_) > 0)) {
```

Comment:
I'm wondering if we were supposed to check the lower bound here? Also, I think `IsVisible` already enforces UB/LB for us, no?

Reply:
> I'm wondering if we were supposed to check the lower bound here?

Are you referring to the user key lower bound? I think it's guaranteed in https://github.com/facebook/rocksdb/blob/7.4.fb/db/db_iter.cc#L737:L744.

Reply:
Oh I think you are referring to timestamp lb.

Reply:
Right, I mean `timestamp_lb_`

Reply:
Thanks for clarification.
The purpose of the check
```cpp
if (!visible &&
        (timestamp_lb_ == nullptr ||
         user_comparator_.CompareTimestamp(ts, *timestamp_ub_) > 0)) {
```
here is to find out if we are done with the current user key (without timestamp).
If it's not visible, then there are two possibilities
- timestamp_lb_ is null, it means the caller just wants the most recent version of the key. Therefore, if it's invisible, it means we have seen a version of this user key greater than the requested timestamp. We are done  since we are iterating from low ts to high ts in Prev()
- timestamp_lb_ is not null. In this case, if it's invisible, we need to distinguish between whether the key's timestamp is too new or too old. If it's too new, we know we are done with this user key. If it's too old, we keep calling Prev() or use FindValueForCurrentKeyUsingSeek().

Reply:
Ah, makes sense. Thanks for the explanation!